### PR TITLE
Designed some test cases of the file 'python-programming-basics.ipynb'

### PR DESCRIPTION
**Pull request by HaoqiTan**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] [local build the book](https://github.com/open-academy/machine-learning/blob/main/CONTRIBUTING.md#build-the-book)
- [x] review the artifacts from local build
- [x] [local build the slides](https://github.com/open-academy/machine-learning/blob/main/CONTRIBUTING.md#build-the-slides-optional) if appliable
- [x] Unit test if appliable

# Checklist:

- [x] My code follows the [style guidelines](https://github.com/open-academy/machine-learning/blob/main/STYLE_GUIDE.md) of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have applied tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by reveurmichael** [comment]
Very good job in general.

One issues: the name of function should follow this rule:
https://stackoverflow.com/questions/159720/what-is-the-naming-convention-in-python-for-variable-and-function

> Function names should be lowercase, with words separated by underscores as necessary to improve readability. Variable names follow the same convention as function names.

File: open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb [hunk 1]
```diff
@@ -1,12 +1,27 @@
 {
```
Comment (reveurmichael) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>the name of function should follow this rule:</p><blockquote><a href="https://stackoverflow.com/questions/159720/what-is-the-naming-convention-in-python-for-variable-and-function" target="_blank">https://stackoverflow.com/questions/159720/what-is-the-naming-convention-in-python-for-variable-and-function</a>Function names should be lowercase, with words separated by underscores as necessary to improve readability. Variable names follow the same convention as function names.</blockquote><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (reveurmichael) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>Assert error. Please revise.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (reveurmichael) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>This cell should be markdown, not code.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

**Comment by reveurmichael** [comment]
Also, before pushing your code, please empty all your output cells.

**Comment by reveurmichael** [comment]
image

File: open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb (hunk 1 above)
Comment (zhangqi444) [reply open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>Just name this as string_lower()</p>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    def string_upper_cased(string):
</pre><p>same as above, <code>string_upper()</code></p><p>Please apply this way to name all your new created functions.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>The class name should be like <code>TestStringUpper()</code></p><p>And you should add some more tests for the unhappy cases in the next PR. Please apply this to all your tests</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.        return string.replace(strpiece1,strpiece2)
</pre><p>Please format all your code by following https://github.com/open-academy/machine-learning/blob/main/STYLE_GUIDE.md#code</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #9.        def test_string_replaced(self):
</pre><p>This is just the happy case. We should have unhappy case later. e.g. to test an empty string, or a string has nothing to replace</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>recover all the code after this cell if you don't want to commit them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>recover the code if you don't want to commit them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='57'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>recover the code if you don't want to commit them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='61'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    # modefied.extend(original[slice(3)])
</pre><p>remove the unused code.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='61'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>recover the code if you don't want to commit them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='72'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>recover the code if you don't want to commit them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='76'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>recover the code if you don't want to commit them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='80'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>recover the code if you don't want to commit them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='85'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <p>recover the code if you don't want to commit them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='89'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    del my_dict["ham"]
</pre><p>recover the code if you don't want to commit them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='93'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    d = my_dict.get('d')
</pre><p>recover the code if you don't want to commit them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='97'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    d = my_dict.get('d')
</pre><p>recover the code if you don't want to commit them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='102'/>

Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.    del my_dict["special_key"]
</pre><p>recover the code if you don't want to commit them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='107'/>

Comment (zhangqi444) [reply open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb]: <p>Did you address this? </p>

Comment (zhangqi444) [reply open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-basics.ipynb]: <p>Please address this comment.</p>

File: open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-introduction.ipynb [hunk 2]
```diff
@@ -100,21 +100,47 @@
   },
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-introduction.ipynb:100]: <p>Such code is not your change. Please exclude them from the PR.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

File: open-machine-learning-jupyter-book/assignments/set-up-env/first-assignment.ipynb [hunk 3]
```diff
@@ -24,22 +24,22 @@
   },
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/set-up-env/first-assignment.ipynb:24]: <p>Such code is not your change. Please exclude them from the PR.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/open-academy/machine-learning/pull/166/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (zhangqi444) [reply open-machine-learning-jupyter-book/assignments/set-up-env/first-assignment.ipynb:24]: <p>Did you resolve this?</p>

File: open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-introduction.ipynb (hunk 2 above)
Comment (zhangqi444) [reply open-machine-learning-jupyter-book/assignments/prerequisites/python-programming-introduction.ipynb:100]: <p>?</p>

**Pull request closed without merging**